### ISSUE TITLE
Qwen3VL 4B and 8B perfomance totally different training curve during fine tune

### ISSUE THREAD

**Issue by siri-rouser** [QwenLM/Qwen3-VL#1885]
## Summary

I am fine-tuning a Qwen3-VL model for a video anomaly classification task using LoRA. I define 5 anomaly categories as new label tokens and train the model to output exactly one of these tokens.

Using the same code, hyperparameters, and data, I observe that:

- Qwen3-VL-4B trains well and learns the classification task.
- Qwen3-VL-8B has much higher loss and appears to learn nothing.

I would like help understanding why the 8B model behaves so differently from the 4B model under the same setup.

---

## Task / Objective

The goal is to fine-tune Qwen3-VL so that it can perform anomaly classification on traffic videos by outputting one of 5 special label tokens:

- <|CAT_NO_ANOMALY|>
- <|CAT_SPEED_TRAJ|>
- <|CAT_DIRECTION_SPACE|>
- <|CAT_CONFLICT_COLLISION|>
- <|CAT_STOP_ROW|>

The model should:

1. Look at the whole video.
2. Choose the single most appropriate category.
3. Output exactly one of the 5 label tokens and nothing else.

---

## Data Format

"""
Example JSON training sample:

<img width="1430" height="285" alt="Image" src="https://github.com/user-attachments/assets/0114bb54-9078-4481-b4d3-21e2dfd6df43" />

"""

Notes:

- I modified _build_message() in data_processor.py to pass through system_prompt.
- Training results are identical whether or not the system prompt is included.

---

## Adding New Label Tokens

Below is the code used to register new tokens and initialize their embeddings:

CATEGORY_LABEL_TOKENS = [
    "<|CAT_NO_ANOMALY|>",
    "<|CAT_SPEED_TRAJ|>",
    "<|CAT_DIRECTION_SPACE|>",
    "<|CAT_CONFLICT_COLLISION|>",
    "<|CAT_STOP_ROW|>",
]

# Add as normal tokens
num_added_tokens = tokenizer.add_tokens(CATEGORY_LABEL_TOKENS)
if hasattr(processor, "tokenizer"):
    processor.tokenizer.add_tokens(CATEGORY_LABEL_TOKENS)

# Resize embeddings
if num_added_tokens > 0:
    model.resize_token_embeddings(len(tokenizer))

embedding_layer = model.get_input_embeddings()
with torch.no_grad():
    base_tokens = ["normal", "speed", "direction", "collision", "stop"]
    for cat_tok, base in zip(CATEGORY_LABEL_TOKENS, base_tokens):
        cat_id = tokenizer.convert_tokens_to_ids(cat_tok)
        base_id = tokenizer(base, add_special_tokens=False).input_ids[0]
        embedding_layer.weight[cat_id].copy_(
            embedding_layer.weight[base_id]
            + torch.randn_like(embedding_layer.weight[base_id]) * 0.01
        )

Both Qwen3-VL-4B and Qwen3-VL-8B are trained using the same code, same LoRA configuration, same dataset, and same hyperparameters.

---

### Qwen3-VL-8B

- Loss starts very high and plateaus around 2.5.
- Curve is smooth but does not decrease.

<img width="606" height="308" alt="Image" src="https://github.com/user-attachments/assets/1aefe5fb-a8d4-4488-9e81-4cf9337a5283" />

### Qwen3-VL-4B

- Loss is less smooth but decreases to around 0.5.
- By epoch 5, loss reaches 0.1–0.3.

<img width="614" height="307" alt="Image" src="https://github.com/user-attachments/assets/23decfb1-07ac-424a-9963-8b59de255fdd" />

---

## Expected vs Actual Loss

Since this is a 5-class classification task:

- A random guess gives expected loss ≈ -log(0.2) ≈ 1.6.

However:

- The 8B model plateaus at ~2.5 (worse than random).
- The 4B model reaches ~0.1–0.3 and performs well.

Evaluation confirms:

- 8B: does not learn the task.
- 4B: learns the task correctly.

This is unexpected given identical setup.

---

## Questions

1. Is there any known issue or additional configuration required when:
   - adding custom tokens to Qwen3-VL, or
   - applying LoRA to Qwen3-VL-8B specifically?

2. Are there recommended best practices for initializing or training new label tokens for Qwen3-VL (especially larger variants)?

3. Could architectural differences explain why the 8B model:
   - stalls at high loss (~2.5),
   - while the 4B model learns normally?

4. What debugging steps would you recommend?
   Examples: verifying token IDs, checking loss masking, validating LoRA target modules, inspecting video feature extraction paths, etc.

---

Thank you for any help or insights.

**Comment by kkontras** [comment]
I would have this question too. It seems like plugging in a CLS head on top of aggregated tokens is quite suboptimal. Let me know if you got somewhere with this.

**Comment by siri-rouser** [comment]
> I would have this question too. It seems like plugging in a CLS head on top of aggregated tokens is quite suboptimal. Let me know if you got somewhere with this.

It spent me so much time and finally I still not figure it out, everything performs randomly after i add some new tokens among the orginal vocabulry. I ended it up with asking model to predict A,B,C,D.

ps: I provide the definition of A,B,C,D through the system prompt.

**Comment by kkontras** [comment]
I see I have tried scraping the correct answer from the response and then using either the last token or any mean pooling to get the cls representation. The scraping led to a non-differentiable answer, so while it's nice for some more understanding on the response, it didn't prove useful. 

On the last/mean token representation, I had some more luck, but it was still overconfident in the wrong ones. Normalizing these output representations or using a trainable CLS token could be an option, but I was hoping to avoid searching for the answer on this one. Probably many will face this.

**Comment by siri-rouser** [comment]
I’m not sure I fully understand your suggestion yet, so please correct me if I’m misinterpreting it.

From what I’ve observed on my side, the model seems to stop learning after I introduce new label tokens, even if i say the loss curve, so I suspect I may be doing something wrong in the way I add or train those tokens.

I tried a few alternatives to avoid “scraping” the answer from free-form text:
1. CLS-style embedding approach: I attempted to represent each class using the embedding of a corresponding token (or a few semantically related base tokens), and treat that as a “CLS embedding” for classification.
2. Constrained generation + probability extraction: I also tried forcing the model to output only one of the label tokens and then reading the softmax probability over just those token IDs.

However, none of these worked reliably in my setup—the training still behaves as if it’s random (loss stays high and doesn’t improve), especially after adding custom tokens.

----
Hope that helps

**Comment by kkontras** [comment]
I would like to edit this response with some further updates I have had. 

I did not understand why you would like to have a <cls> token for each class, but having one for all is a good idea. I did that with a learnable one and made sure to unfreeze the embeddings, then added a hook to mask all the gradients except for the cls one. I unfroze the norms, too, but I am not sure if this has any severe effect. Then, on top of the <cls>, I trained a linear classifier, and it got decent results. The loss so far doesnt seem to be minimized, but that can be expected, I guess, since I train only the token and the linear head. I will probably proceed into STF or Lora. I will post the results then. 

```        lm = qwen.model.language_model
            lm.embed_tokens.weight.requires_grad = True
            cls_id = int(self.cls_token_id)
            mask = torch.zeros_like(emb.weight, dtype=torch.float32)
            mask[cls_id].fill_(1.0)
            def grad_mask_hook(grad):
                return grad * mask.to(grad.device, grad.dtype)
            if not hasattr(self, "_cls_grad_hooked"):
                lm.embed_tokens.weight.register_hook(grad_mask_hook)
                self._cls_grad_hooked = True
```

**Comment by kkontras** [comment]
I have worked so far only with the 2B, and I finetuned in one case only the cls token plus the classifier, and in a second case, LoRA over the LLM with the cls and classifier. Working only with the cls in ScienceQA returns 40% acc, while with LoRA finetuning can rise above 75%. 

I am not sure how I can help further, but feel free to ask anything that helps.
